### PR TITLE
[FLINK-30128] Introduce Azure Data Lake Gen2 APIs in the Hadoop Recoverable path

### PR DESCRIPTION
**Pull request by ramkrish86**
## What is the purpose of the change

Supports ABFS as Streaming file sink

## Brief change log

- Allows sink to write to ABFS
- Allows to recover from ABFS
- There is no truncate API in ABFS hadoop driver. But for cases where we want to recover a job from an older persisted offset, we need to truncate. For which we rewrite the file.

## Verifying this change

This change added tests and can be verified as follows:
  - Verified the basics in a Kubernetes based session cluster with 1 JM and 4 TMs. 
  - Was able to write to ABFS and also recover from it. 
  - **Pls note this is an initial PR, we are still doing some extensive testing. This is for initial feedback**
  -  Used the existing test cases in **HadoopRecoverableWriterTest** and extended the same to verify the functionality but with ABFS.
  - New test case name AzureBlobRecoverableWriterTest
  - Existing Hadoop serializers are used. Nothing changed in this format.
  - `[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.flink.fs.azurefs.AzureBlobRecoverableWriterTest
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 72.345 s - in org.apache.flink.fs.azurefs.AzureBlobRecoverableWriterTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (No)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes) - Recovery is also handled here.
  - The S3 file system connector: (No)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)
-->

### CODE REVIEW

**Comment by carp84** [comment]
Thanks for the contribution @ramkrish86 .

@xintongsong could you help review this one since you have experience reviewing similar support for Google storage FS (PR#15599)? Many thanks.

**Comment by xintongsong** [comment]
Sure. I'll try to take a look asap.

**Comment by ramkrish86** [comment]
@xintongsong  - A gentle ping!!!

File: flink-filesystems/flink-azure-fs-hadoop/pom.xml
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/pom.xml:194]: Just trying to understand, why do we need this change?

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileSystem.java
Comment (xintongsong) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileSystem.java:40]: This change is unnecessary. You can use `getHadoopFileSystem()` where it's needed.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java:102]:
```suggestion
```

Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java:102]: And then we won't need the visibility change for `HadoopRecoverableSerializer`.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java:71]:
```suggestion
```

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFileSystem.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFileSystem.java:38]: Can be package private.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:44]: This class is very similar to `HadoopRecoverableFsDataOutputStream`. The only differences are in the initialization and the committer implementation. I wonder if they can be further deduplicated. I'm particularly interested in hiding `HadoopFsRecoverable` from this submodule.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:63]: Can be package private.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:73]: Can be package private.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:124]: Better to check the actual number of bytes being read.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:129]: 2 placeholders while only 1 argument is provided.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: This means that, despite the src and dest files are on the same remote file system, all the data being copied have to go through the local system where this piece of code is executed. I wonder if there's a better way doing this. (Maybe not, just trying to double-check on this.)

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:285]: I think we should explicitly throw an exception here, because the data cannot be committed.

**Review by xintongsong — COMMENTED** [review]
@ramkrish86, thank for the contribution. I have left some comments, please take a look.

In addition to the inline comment:
- The CI failed due to some code formatting issues. See [this page](https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting) for setting up formatting tools in your IDE.
- There're many `TODO`s in `AzureBlobFsRecoverableDataOutputStream`. What's your plan for them?

File: flink-filesystems/flink-azure-fs-hadoop/pom.xml
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/pom.xml:194]: Since currently we are extending these HadoopRecoverable, this not being in the jar while I submit the job it is not able to pick it up. This was the initial problem that users were reporting too.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java:102]: Got it sure.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFileSystem.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFileSystem.java:38]: Ok sure . will change it.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:73]: Sure will change it.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:129]: Done. Great catch.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:124]: Yes. Will update it. Its a good way to handle EOF. Also am now adding 'close' of the input stream. It might not be a critical one but better to do it.

**Comment by ramkrish86** [comment]
> * There're many `TODO`s in `AzureBlobFsRecoverableDataOutputStream`. What's your plan for them?

Mainly 2 TODOs. One is the 2MB buffer and next is the flush support. I think for now it should be good. But I can raise TODO JIRAs and work on it later. Is that ok ?

**Comment by ramkrish86** [comment]
Pushed the changes. We will update here once the other tests are done.

**Comment by ramkrish86** [comment]
@flinkbot run azure

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: I meant the way we write the truncated content to the renamed temp file. We first read the data from the temp file into the memory, then write the data from memory to the renamed temp file. Here the memory is where the Flink file sink operator is executed, while both temp file and renamed temp file are on the remote ABFS. Data transmissions between the memory and the ABFS can be expensive, at least more expensive than data transmissions within ABFS. So I just wonder if there's anyway to copy data directly from the temp file to the renamed temp file, without going through the local memory.

Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: Got your question. I checked with the storage driver team (hadoop based storage driver), they said currently there is no direct truncate support. So the manual truncation is the option over here. A generic question though @xintongsong - In case of ORC/Parquet format does the file recovery flow take care of the truncate or those formats already are aware of the last committed offset? This is a question out of this ABFS support PR (in general)

File: flink-filesystems/flink-azure-fs-hadoop/pom.xml
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/pom.xml:194]: Can I resolve this?

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: Bulk-encoded formats (ORC, Parquet, etc.) do not need truncation, because in-progress files are guaranteed to be finished / flushed on checkpoints. Truncation is only needed when an in-progress file can live through multiple checkpoints, which is only supported for row-encoded formats.

File: flink-filesystems/flink-azure-fs-hadoop/pom.xml
Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/pom.xml:194]: Sure. Resolved.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: Thanks @xintongsong  for this info.

Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: Can we resolve this conversation also @xintongsong - How ever after our tests I can update here before we want to commit this - if that is fine with you.

Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:113]: Sounds like there's no off-the-shelf solution for this at the moment. I think we should at least understand the performance impact and mention it in the user documentation.

Let's leave this conversation unresolved as a reminder for now. This won't block the PR review.

**Comment by ramkrish86** [comment]
Found an issue with 3.3.2 and above ABFS version. The test case I made it work functionally by changing the hadoop version to 3.3.1 and it works fine. A minor config update might be needed. Will test and confirm  back .

**Comment by ramkrish86** [comment]
The current failure seems to be unrelated.

**Comment by xintongsong** [comment]
@flinkbot run azure

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (anoopsjohn) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:163]: What if there is a fail in between these steps of delete and rename?  Later how it will be possible to rollback.  The truncate should be able to handle such case?

Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:163]: In HDFS case we try to trucnate and if that is still failing we were throwing exceptions. So I think the same is happening here. This recovery will be a failure. Also as I said truncate happens only in case where we move to an older recovery checkpoint.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:115]: Not sure. May be in testing if this case happens will handle it of removing the *.rename file once.

**Comment by xintongsong** [comment]
Hi @ramkrish86,

How are things going? I briefly checked the PR and it seems most of the comments are addressed, except for the one about deduplicating `HadoopRecoverableFsDataOutputStream` and `AzureBlobFsRecoverableDataOutputStream`. Please let me know if you're ready for another round of review.

BTW, do you need this feature to be included in Flink 1.17? If yes, the feature freeze date for 1.17 is Jan 31st, 4 weeks from now. However, I have a vacation plan and won't be responsible in the last 2 weeks of the 4. So we probably would need to finalize this PR, as well as updating the documentations, in the next 2 weeks. If there's no such a strong demand to see this in 1.17, take your time then.

**Comment by ramkrish86** [comment]
@xintongsong  - We have almost completed the testing part also. I will update the PR in a day or two time frame. Am checking with other comments that Anoop had added over here.

**Comment by ramkrish86** [comment]
>  except for the one about deduplicating HadoopRecoverableFsDataOutputStream and AzureBlobFsRecoverableDataOutputStream
What specifically you are talking about here? I went through the older comments and I thought I resolved them which ever was discussed.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:44]: > What specifically you are talking about here?

@ramkrish86, I meant this one. I guess it's overlooked because github hides it when there's too many comments.

Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java]: I have tried refactoring this . I am not able to remove HadoopFsRecoverable totally as it is also getting passed in the constructor. But avoided some duplication between HadoopRecoverableFsDataOutputStream and AzureBlobFsRecoverableDataOutputStream

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:106]: Can we resolve this if rename by default is with 'overwrite'?

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:154]: In our testing with DefaultRollingpolicy - the recovery is not really possible as the inprogress that was marked would have already been committed in the next checkpoint. Only with policy as Checkpoint based rolling policy we can achieve this . So basically when moving to an older checkpoint - I believe  yes if we fail in a delete and rename step it is a non recoverable error but we can always move to latest checkpoint. But am not very sure how this recovery will be used in case of FS based sinks.

Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java:185]: Now I have handled this case by checking if the '.rename' file is available and if that is really of the expected length just go with rename. This should solve the problem of where the failure happens between delete and rename.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (ramkrish86) [reply flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java]: @xintongsong  - Can you pls check if this refactoring is fine.

**Comment by ramkrish86** [comment]
@xintongsong  - We are done with the testing too. Will wait for your final set of comments. Thanks.

**Comment by xintongsong** [comment]
@ramkrish86, thanks for the notice. I'll try to find time for this asap.

**Comment by ramkrish86** [comment]
Thanks @xintongsong . Also it would be great if this can be there in 1.17 release.

**Comment by ramkrish86** [comment]
We used the streaming wordcount program as the test case. 
We generated close to 2G data and collected the word count sample. This result was matched by using output as ABFS file sink and verified the output by running the tests without restarting of any TMs and also by restarting of TMs intermittently to ensure we are able to recover and get back the same word count.  Total TMs - 5. Parallelism - 4. 

We tested with default rolling policy where we are sure the inprogress files will be created and it needs a truncation on recovery from the latest check point. Also tested with checkpoint based rolling policy and in that case no truncation was needed as inprogress files were always committed. 
We have not extensively tested with ORC/Parquet formats. 
I can commit for any other further improvements/enhancements/bug fixes that needs to be done here. 

Created the TODO as agreed up on. https://issues.apache.org/jira/browse/FLINK-30588. FYI @xintongsong

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/BaseHadoopFsRecoverableFsDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/BaseHadoopFsRecoverableFsDataOutputStream.java]: 1. IIUC, @anoopsjohn 's comment was that you can call `hsync` without calling `hflush` in `sync()` for ABFS. It was not about ABFS should call `hsync` in `flush()`. Though the two methods do the same thing, according to him.
2. Not needing to call `hflush` before `hsync` sounds to be a specialty of ABFS. We probably should make `AzureBlobFsRecoverableDataOutputStream` override the `flush/sync` method, rather than make the method abstract in the base class.

File: flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java
Comment (xintongsong) [inline flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java]: We use `@VisibleForTesting` for fields that have broader visibility then needed for testing purpose.

**Review by xintongsong — APPROVED** [review]
@ramkrish86, thanks for addressing the comments, and for the thorough tests. The PR LGTM. Taking over from here.

I have squashed the commits, updated the commit message and rebased the latest master branch. I also have a few minor comments, which I addressed myself in a fixup commit. Please take a look.

If there's no objections, I'll merge this PR after the CI passes.

**Comment by ramkrish86** [comment]
@xintongsong  - Thanks for the approval. Heartfelt thanks for shepherding and helping with the PR reviews. 
BTW want to bring to your notice that before raising this PR I had asked a query in the dev@ about running the Azure tests in the CI. I was referred to https://cwiki.apache.org/confluence/display/FLINK/Continuous+Integration and also was told to contact Martjin. Seems the Azure IT tests runs only on CI for master. So may be after intergration we might have to check if the tests really run in CI. However I have ensured the tests  have passed successfully in my local.

**Comment by xintongsong** [comment]
@ramkrish86,
It was nice to cooperate with you. Thanks for the contribution.

Concerning the IT tests, I think they are only available to the cron builds for master and release branches (like [this](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=44569&view=results)). We have release managers monitoring the the cron builds and reporting issues to jira. I'll also pay attention to this in the first couple of days after merging this.

Just to double check. I thought we also need to update the user documentation for this change. However, I just checked the documentation and it seems never mentioned that `abfs://` doesn't work with the file sink. So nothing needs to be changed for the documentation, right?

**Comment by ramkrish86** [comment]
https://nightlies.apache.org/flink/flink-docs-release-1.13/docs/connectors/datastream/streamfile_sink/#important-considerations
Here we say that we support S3, HDFS and local. Probably we should add ABFS.

**Comment by ramkrish86** [comment]
@xintongsong  - Also want to know about this https://nightlies.apache.org/flink/flink-table-store-docs-master/docs/filesystems/overview/.
Now since ABFS support is available in FileSink (after this PR gets merged), shall we have the similar support for table store too?

**Comment by xintongsong** [comment]
> https://nightlies.apache.org/flink/flink-docs-release-1.13/docs/connectors/datastream/streamfile_sink/#important-considerations Here we say that we support S3, HDFS and local. Probably we should add ABFS.

Nice catch. I guess we should add both ABFS and WASB here.

**Comment by xintongsong** [comment]
> @xintongsong - Also want to know about this https://nightlies.apache.org/flink/flink-table-store-docs-master/docs/filesystems/overview/. Now since ABFS support is available in FileSink (after this PR gets merged), shall we have the similar support for table store too?

Good question. From the documentation, it seems not all file systems that FilsSink supports are supported by FTS. I'll talk to someone from the FTS team about this.

**Comment by xintongsong** [comment]
> @xintongsong - Also want to know about this https://nightlies.apache.org/flink/flink-table-store-docs-master/docs/filesystems/overview/. Now since ABFS support is available in FileSink (after this PR gets merged), shall we have the similar support for table store too?

Checked with @JingsongLi about this. Currently, when using Flink + FTS, all file systems in that flink supports are supported. However, when using other FTS with other engines (Spark / Hive / Trino), further development is required on the FTS side, mostly related to dependency bundling and conflict resolving.

That is to say, with what we currently have, there's no need to update the FTS documentation. And if we want to support FTS + ABFS + Spark/Hive/Trino, we would need to open another PR to the FTS project, which is definitely welcome.

**Comment by ramkrish86** [comment]
> That is to say, with what we currently have, there's no need to update the FTS documentation. And if we want to support FTS + ABFS + Spark/Hive/Trino, we would need to open another PR to the FTS project, which is definitely welcome.

Perfect. Will keep a watch on this and we will raise PR accordingly. Thanks once again @xintongsong

**Pull request closed without merging**
